doc/mpls/MPLS_hld.md

```diff
@@ -0,0 +1,563 @@
+
+``` rfc5234
+INTERFACE|{{ifname}}
+    "mpls":{{enable|disable}} (OPTIONAL)
+
+; Defines schema for MPLS configuration attribute
+key          = INTERFACE:ifname    ; Interface name
+; value annotations
+ifname       = 1*64VCHAR           ; name of the Interface
+; field      = value
+mpls         = "enable"/"disable"  ; Enable/disable MPLS function. Default "disable"
+```
+
+##### PORTCHANNEL_INTERFACE
+The existing PORTCHANNEL_INTERFACE table is enhanced to accept a new "mpls" enable/disable attribute.
+
+``` rfc5234
+PORTCHANNEL_INTERFACE|{{ifname}}
+    "mpls":{{enable|disable}} (OPTIONAL)
+
+; Defines schema for MPLS configuration attributes
+key         = PORTCHANNEL_INTERFACE:ifname  ; Port Channel Interface name
+;value annotations
+ifname      = 1*64VCHAR                     ; name of the Interface (Port Channel)
+; field     = value
+mpls        = "enable"/"disable"            ; Enable/disable MPLS function. Default "disable"
+```
+
+##### VLAN_INTERFACE
+The existing VLAN_INTERFACE table is enhanced to accept a new "mpls" enable/disable attribute.
+
+``` rfc5234
+VLAN_INTERFACE|{{ifname}}
+    "mpls":{{enable|disable}} (OPTIONAL)
+
+; Defines schema for MPLS configuration attributes
+key         = VLAN_INTERFACE:ifname          ; VLAN Interface name
+;value annotations
+ifname      = 1*64VCHAR                      ; name of the Interface (VLAN)
+; field     = value
+mpls        = "enable"/"disable"             ; Enable/disable MPLS function. Default "disable"
+```
+##### CRM Config
+The existing CRM Config stanza is enhanced to include new MPLS CRM attributes.  These attributes parallel existing CRM configuration for other resource types (eg, IPv4/IPv6 routes).
+
+``` rfc5234
+CRM
+  Config
+    "mpls_inseg_threshold_type":{{percentage|used|free}} (OPTIONAL)
+    "mpls_inseg_high_threshold":{{UINT32}} (OPTIONAL)
+    "mpls_inseg_low_threshold":{{UINT32}} (OPTIONAL)
+
+; Defines schema for CRM MPLS configuration attributes
+; field                    = value
+mpls_inseg_threshold_type  = "percentage"/"used"/"free" ; Threshold type. Default "percentage"
+mpls_inseg_high_threshold  = UINT32             ; High threshold. Default value = 85
+mpls_inseg_low_threshold   = UINT32             ; Low threshold. Default value = 70
+```
+
+#### YANG Model Enhancements
+The existing sonic-interface.yang model is enhanced to support a new "mpls" enable/disable attribute.
+
+``` rfc5234
+  container sonic-interface {
+    container INTERFACE {
+      list INTERFACE_LIST {
++       leaf mpls {
++         type string {
++           pattern "enable|disable";
++         }
++       }
+      }
+    }
+  }
+```
+
+### Warmboot and Fastboot Design Impact
+MPLS design will not affect warmboot or fastboot design.
+
+### Restrictions/Limitations
+- No support for programming default MPLS routes for IPv4/IPv6 Explicit NULL from SONiC infrastructure. These default MPLS routes must be programmed/handled by vendor SAI implementation.
```

Comment:
If there is no default entry for V4/V6 explicit null, how is VRF determined for the packet? In SAI, explicit NULL MPLS label can point to MPLS RIF and that can determine the VRF.